core/desugarDatatypes.ml

```diff
@@ -111,6 +118,31 @@ exception UnexpectedOperationEffects of string
 
 module Desugar =
 struct
+  (* Desugars quantifiers into Types.quantifiers,
+   * returning updated variable environment.
+   * Lifted / deduplicated from `typename` and `Forall` desugaring. *)
+  let desugar_quantifiers (var_env: var_env) (qs: Sugartypes.quantifier list) :
+      (Types.quantifier list * var_env) =
+      ListLabels.fold_right ~init:([], var_env) qs
+      ~f:(fun q (args, {tenv=tenv; renv=renv; penv=penv}) ->
```

Comment:
Shorten this code by using field punning and avoiding repetition between the branches:

```OCaml
      ~f:(fun (name, (primarykind, subkind), _freedom)
              (args, {tenv; renv; penv}) ->
            let var = Types.fresh_raw_variable () in
            let subkind = concrete_subkind subkind in
            match primarykind with
              | PrimaryKind.Type ->
                  let point = Unionfind.fresh (`Var (var, subkind, `Rigid)) in
                    ((var, subkind, `Type point)::args,
                     {tenv=StringMap.add name point tenv; renv; penv})
              | PrimaryKind.Row ->
                  let point = Unionfind.fresh (`Var (var, subkind, `Rigid)) in
                    ((var, subkind, `Row point)::args,
                     {tenv; renv=StringMap.add name point renv; penv})
              | PrimaryKind.Presence ->
                  let point = Unionfind.fresh (`Var (var, subkind, `Rigid)) in
                    ((var, subkind, `Presence point))::args,
                     {tenv; renv; penv=StringMap.add name point penv})
```
Is there a trick to float `point` binding out of the branches? The expression is the same in all the three branches but it has a different type in each of them.

Reply:
I couldn't find a way to float `point` out, unfortunately. Thanks for your comments in shortening the code, though, which I've added.